f--
title: Exporatory genetics of HpA
date: 2020-11-20
author: Kevin Murray
---

```{r setup}
if (!require("tidyverse"))    { install.packages("tidyverse")     ; require("tidyverse")    }
if (!require("foreach"))      { install.packages("foreach")       ; require("foreach")      }
if (!require("doParallel"))   { install.packages("doParallel")    ; require("doParallel")   }
if (!require("parallel"))     { install.packages("parallel")      ; require("parallel")     }
if (!require("ggplot2"))      { install.packages("ggplot2")       ; require("ggplot2")      }
if (!require("ggrepel"))      { install.packages("ggrepel")       ; require("ggrepel")      }
if (!require("ggmap"))        { install.packages("ggmap")         ; require("ggmap")        }
if (!require("RColorBrewer")) { install.packages("RColorBrewer")  ; require("RColorBrewer") }
if (!require("fossil"))       { install.packages("fossil")        ; require("fossil")       }
if (!require("vegan"))        { install.packages("vegan")         ; require("vegan")        }
if (!require("gdm"))          { install.packages("gdm")           ; require("gdm")          }
if (!require("ecodist"))      { install.packages("ecodist")       ; require("ecodist")      }
if (!require("SNPRelate"))    { BiocManager::install("SNPRelate") ; require("SNPRelate")    }
if (file.exists("data/01_hpa-geogenetics.Rda")) load(file="data/01_hpa-geogenetics.Rda")
if (!dir.exists("out/plot")) dir.create("out/plot", recursive=T)
```

# Genotype data

```{r makegds}
if (!file.exists("data/HaR.filtered_snps_final.PASS.bi.hardFiltered.indFiltered.noMit.reheader.gds")) {
    cat("Updating GDS file from VCF\n")
    snpgdsVCF2GDS("data/HaR.filtered_snps_final.PASS.bi.hardFiltered.indFiltered.noMit.reheader.vcf.gz",
                  "data/HaR.filtered_snps_final.PASS.bi.hardFiltered.indFiltered.noMit.reheader.gds")
} else {
    cat("Not updating gds file\n")
}
```


```{r gdsopen, cache=F}
gds = snpgdsOpen("data/HaR.filtered_snps_final.PASS.bi.hardFiltered.indFiltered.noMit.reheader.gds", allow.duplicate=T)
gds.sum  = snpgdsSummary(gds)
```


## Broad data exploration: missingness, freqs

### Sample missingness

```{r sample.missing}
samp.miss = snpgdsSampMissRate(gds)
hist(samp.miss, main="Sample missing prop")
```

Looks broadly good, about three crappy samples with < 80% completeness. Will go
ahead with those in there though as they still look reasonable. If they stick
out in the dendro then I'll remove for further work. Might be worth having a go
with imputation.

### SNP missingness

```{r snp.rate}
snp.rate = snpgdsSNPRateFreq(gds)
hist(snp.rate$MissingRate, main="SNP missing prop")
```

Again per-snp missingness is very good. Will set the snp missing threshold at
30% as that removes the worst `r mean(snp.rate$MissingRate>0.30)*100`% of SNPs

### SNP MAF

```{r snp.rate.hist}
hist(snp.rate$AlleleFreq, main="SNP AF")
hist(snp.rate$MinorFreq, main="SNP MAF")
```

Interesting SFS. Fairly typical folded SFS, but it seems the reference allele
is frequently the rarer allele.


### Set thresholds

```{r set.thresh}
sample.missing.thresh = 0.2
snp.missing.thresh = 0.3
```


## Dendrogram/PCA

```{r snp.ibs}
snp.ibs = snpgdsIBS(gds, missing.rate = snp.missing.thresh, num.thread = 4)
image(1-snp.ibs$ibs)
```

```{r dendro}
snp.ibs %>%
    snpgdsHCluster() %>%
    snpgdsCutTree() %>%
    snpgdsDrawTree()
```

```{r pca}
snp.pca = snpgdsPCA(gds, missing.rate = snp.missing.thresh)
plot(snp.pca)
```


# Geographic exploration

## Read metadata,

```{r geometa}
geo = read_csv("data/finalInds.AllGeoInfo.csv")
summary(geo[, c("latitude", "longitude")])
```

## Maps

```{r basemap}
eu.bbox = c(left=-18, right=32, top=62, bottom=32)
basemap.terr = get_stamenmap(eu.bbox, zoom=5, maptype="terrain-background")
p =  ggmap(basemap.terr, darken=c(0.4, "white"), legend="topleft", extent="device")
print(p)
```

```{r makemap}
eu.geo = geo %>%
    filter(latitude > eu.bbox["bottom"], latitude < eu.bbox["top"], longitude > eu.bbox["left"], longitude < eu.bbox["right"])
samp.within.eur = eu.geo %>%
    pull(ind)
```

```{r mapplot}
p + geom_point(aes(longitude, latitude, colour=pop), data=eu.geo,
           alpha=1, size=2) +
    labs(x="Longitude", y="Latitude") +
    scale_colour_discrete(name="Population") + 
    theme_bw()
ggsave("out/plot/01_samples-within-europe-map.svg")
```


## Align to (EU) geno data

Now match the rows/cols of the IBS matrix to get our distance

```{r geoaln}
m = match(samp.within.eur,  snp.ibs$sample.id)
geno.dist = 1 - snp.ibs$ibs[m, m]
dim(geno.dist)
length(geo$ind)
rownames(geno.dist) = colnames(geno.dist) = samp.within.eur
hist(geno.dist)
```

## Simple IBD

```{r ibd}
geo.dist = as.matrix(eu.geo[, c("longitude", "latitude")]) %>%
    earth.dist() %>%
    as.dist()
ibd.geno.dist = geno.dist %>%
    as.dist()
vegan::mantel(ibd.geno.dist, geo.dist)
mg = ecodist::mgram(ibd.geno.dist, geo.dist)
mg$mgram[,4]
plot(mg, xlab="Distance (km)")
```

## GDM

```{r gdm.prep}
gdm.geno = cbind(ind=rownames(geno.dist), as.data.frame(as.matrix(geno.dist)))
gdm.preds = eu.geo %>%
    dplyr::select(ind, latitude, longitude)
gdm.geodist = cbind(ind=eu.geo$ind, as.data.frame(as.matrix(geo.dist)))
dim(gdm.geno)
dim(gdm.preds)
dim(gdm.geodist)
```

```{r formatsitepair}
sp = formatsitepair(gdm.geno, 3, predData=gdm.preds, siteColumn="ind",
                    XColumn="longitude", YColumn="latitude",
                    distPreds=list("geography.gcd"=gdm.geodist)) %>%
    filter(s2.matrix_1 > 5) # only include site pairs with a geographic distance > 5km
mdl = gdm(sp, geo=F)
plot(mdl, include.rug=T, rug.sitepair=sp)
```

# Finalise

```{r cleanup}
save.image(file="data/01_hpa-geogenetics.Rda", compress="xz")
save(sample.missing.thresh, snp.missing.thresh, samp.within.eur,
     file="data/sample_selection.Rda", compress="xz")
```
